### ISSUE TITLE
Add ldap bind user

### ISSUE THREAD

**Issue by t0xicCode** [auraphp/Aura.Auth#45]
It is considered good practice to have ldap authentication libraries first bind using a service account, then run a search for the user and finally try to bind (and thus authenticate) using the provided credentials.

As an added bonus, that would enable supports for multi-level user trees, which the current implementation does not support.

**Comment by pmjones** [comment]
I am all for that. If you have examples in code, or a PR, I'd be happy to a review.

**Comment by t0xicCode** [comment]
Unfortunately I do not have any example readily available. I do know that
the Drupal ldap module supports this use case.

**Comment by pmjones** [comment]
Link to the Drupal LDAP module? Every little bit you help out helps this go faster. :-)

**Comment by t0xicCode** [comment]
I'm on mobile right now, but I'll add some details as soon as I get to a
computer :smiley:

**Comment by harikt** [comment]
@t0xicCode is it simple ldap or ldap ?

http://cgit.drupalcode.org/simple_ldap/tree/?h=7.x-2.x
http://cgit.drupalcode.org/ldap/tree/?h=8.x-2.x

May be good if you can point to the right file though. 

Thanks

**Comment by t0xicCode** [comment]
@harikt ldap. It's located in the function at http://cgit.drupalcode.org/ldap/tree/ldap_authentication/ldap_authentication.inc#n532.

Line [562](http://cgit.drupalcode.org/ldap/tree/ldap_authentication/ldap_authentication.inc#n562) does the initial bind with the service account credentials, it then maps or searches for the given username, and finally, at line [659](http://cgit.drupalcode.org/ldap/tree/ldap_authentication/ldap_authentication.inc#n659) actually authenticates with the given password and the mapped ldap user.

The [`bind` function](http://cgit.drupalcode.org/ldap/tree/ldap_servers/LdapServer.class.php#n322) that is called at multiple places uses the stored service account information if it's passed `NULL` for its parameters.

**Comment by harikt** [comment]
Thank you @t0xicCode

**Comment by enygma** [comment]
There's also some trickiness here that some LDAP servers require you to set options for the bind to even work right: 

```
ldap_set_option($ldap, LDAP_OPT_REFERRALS, 0);
ldap_set_option($ldap, LDAP_OPT_PROTOCOL_VERSION, 3);
```

I'm not sure this breaks it for other connection types, but I had to use this for a MS-based domain handler.

Also, I noticed in the VerifierInterface, it requires the second parameter (the hashed value). In this case, it doesn't make sense to have to generate something for that as the plain-text password is just sent to the LDAP server for validation.

**Comment by harikt** [comment]
> Also, I noticed in the VerifierInterface, it requires the second parameter (the hashed value). In this case, it doesn't make sense to have to generate something for that as the plain-text password is just sent to the LDAP server for validation.

Hm, in that case the verifier don't need to do the verification . And even though there need a PlainText verifier.

**Comment by harikt** [comment]
Sorry forgot to add . Which always returns true.

``` php
<?php
namespace Aura\Auth\Verifier;

class PlainTextVerifier implements VerifierInterface
{
    public function verify($plaintext, $hashvalue, array $extra = array())
    {
        return $plaintext === $hashvalue;
    }
}
```

**Comment by enygma** [comment]
So you're saying not having a verifier for LDAP at all? Based on the others it kind of seems like that's the point (to abstract that out behind the generic `verify` method).

**Comment by harikt** [comment]
@enygma I was looking at http://php.net/manual/en/function.ldap-compare.php and it seems to me we need to do something like as shown in example. The above example I mentioned checking plain and hash one is wrong.

**Comment by enygma** [comment]
Yeah, the tricky part on that is whether or not the password is returned and how to set up that base DN information. Also, the compare will only work if there's a service account as was mentioned in the initial comment here. If you're using the `ldap_bind` method for testing the login, you wouldn't even be able to run the compare if the login was incorrect.

**Comment by harikt** [comment]
New feature in https://github.com/auraphp/Aura.Auth/pull/109

What it does

LdapAdapter now accepts an optional search configuration. When provided, instead of binding the user directly it:

1. binds with a service account,
2. searches for the user to discover their real DN — which can live anywhere in the tree, so users spread across multiple OUs are now supported (the "added bonus" @t0xicCode asked for), and
3. rebinds as that DN to verify the password, returning the user's attributes.

Without a search config it behaves exactly as before (direct bind via dnformat), so it's fully backward compatible.

```
$ldap_adapter = $auth_factory->newLdapAdapter(
    'ldaps://ldap.example.com:636',
    'uid=%s,ou=users,dc=example,dc=org', // ignored when 'search' is given
    [
        LDAP_OPT_PROTOCOL_VERSION => 3,
        LDAP_OPT_REFERRALS => 0,          // the AD/referrals point @enygma raised
    ],
    [
        'binddn'     => 'cn=service,dc=example,dc=org',
        'bindpw'     => 'service-account-password',
        'basedn'     => 'dc=example,dc=org',
        'filter'     => '(uid=%s)',        // %s = escaped username
        'attributes' => ['cn', 'mail'],    // optional; default: all
    ]
);
```

On success the login returns [$username, $attributes]. Error paths are typed: SearchFailed (new — an operational search failure, kept distinct from a genuine zero-result search, which stays UsernameNotFound), MultipleMatches, and BindFailed.

Re: the earlier VerifierInterface/PlainTextVerifier tangent — no verifier is involved; the password is validated by the rebind, exactly as expected for LDAP.

Testing

Beyond the mocked unit tests, there's now a real-OpenLDAP integration suite that runs in CI against a live slapd (installed from apt — no third-party image). It seeds users in two different OUs and verifies both the direct bind and the search flow discovering users across OUs, plus the wrong-password / not-found / search-failure paths.

Docs: see docs/adapters.md (with docs/di-configuration.md and docs/service-idioms.md updated too).

Thanks @t0xicCode for the Drupal reference and @enygma for the AD/options notes — both fed directly into the implementation.

Thanks to Claude and Coderabbit for their generous give away to opensource tools.
